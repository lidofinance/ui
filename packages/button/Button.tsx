import React, { ForwardedRef, forwardRef } from 'react'
import {
  ButtonStyle,
  ButtonContentStyle,
  ButtonLoaderStyle,
} from './ButtonStyles'
import { ButtonProps } from './types'
import { useRipple } from './useRipple'

const loaderSizes = {
  xxs: 'small',
  xs: 'small',
  sm: 'small',
  md: 'medium',
  lg: 'medium',
} as const

function Button(props: ButtonProps, ref?: ForwardedRef<HTMLButtonElement>) {
  const {
    size = 'md',
    variant = 'filled',
    color = 'primary',
    square = false,
    fullwidth = false,
    loading = false,
    active = false,
    onClick,
    disabled,
    children,
    ...rest
  } = props

  const { handleClick, ripple } = useRipple(props)
  const loaderSize = loaderSizes[size]

  return (
    <ButtonStyle
      $size={size}
      $variant={variant}
      $fullwidth={fullwidth}
      $color={color}
      $square={square}
      $loading={loading}
      $active={active}
      onClick={handleClick}
      disabled={disabled || loading}
      type='button'
      ref={ref}
      {...rest}
    >
      <ButtonContentStyle $hidden={loading}>{children}</ButtonContentStyle>
      {loading && <ButtonLoaderStyle size={loaderSize} />}
      {!active && ripple}
    </ButtonStyle>
  )
}

export default forwardRef(Button)
