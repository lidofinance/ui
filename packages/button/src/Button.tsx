import React, { ForwardedRef, forwardRef } from 'react'
import {
  ButtonStyle,
  ButtonContentStyle,
  ButtonLoaderStyle,
} from './ButtonStyles'
import { ButtonButtonProps, ButtonProps } from './types'
import { useRipple } from './useRipple'

const loaderSizes = {
  xxs: 'small',
  xs: 'small',
  sm: 'small',
  md: 'medium',
  lg: 'medium',
} as const

function Button(
  props: ButtonProps,
  ref?: ForwardedRef<HTMLButtonElement | HTMLAnchorElement>
) {
  const {
    size = 'md',
    variant = 'filled',
    color = 'primary',
    square = false,
    fullwidth = false,
    loading = false,
    active = false,
    onClick,
    disabled,
    children,
    href,
    ...rest
  } = props as ButtonButtonProps

  const { handleClick, ripple } = useRipple(props)
  const loaderSize = loaderSizes[size]

  return (
    <ButtonStyle
      as={href ? ('a' as React.ElementType) : undefined}
      $size={size}
      $variant={variant}
      $fullwidth={fullwidth}
      $color={color}
      $square={square}
      $loading={loading}
      $active={active}
      onClick={handleClick}
      disabled={disabled || loading}
      type='button'
      ref={ref}
      href={href}
      {...rest}
    >
      <ButtonContentStyle $hidden={loading}>{children}</ButtonContentStyle>
      {loading && <ButtonLoaderStyle size={loaderSize} />}
      {!active && ripple}
    </ButtonStyle>
  )
}

export default forwardRef(Button)
