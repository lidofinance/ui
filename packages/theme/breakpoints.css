@custom-media --lido-media-breakpoint-up-sm screen and (min-width: 359px);
@custom-media --lido-media-breakpoint-up-md screen and (min-width: 479px);
@custom-media --lido-media-breakpoint-up-lg screen and (min-width: 767px);
@custom-media --lido-media-breakpoint-up-xl screen and (min-width: 1023px);
