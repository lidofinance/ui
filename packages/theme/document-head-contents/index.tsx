import React, { FC } from 'react'
import { Fonts } from './element-fonts'
import { initGlobalColors, StyleThemeColors } from './element-theme-colors'
import { initGlobalCookieTheme, ScriptThemeValue } from './element-theme-script'
export { Fonts } from './element-fonts'

export const initColors = () => {
  initGlobalColors()
  initGlobalCookieTheme()
}

export const LidoUIHead: FC<{
  cssStyleVars?: boolean
  jsStyleScript?: boolean
  cssFont?: boolean
}> = ({ cssStyleVars = true, cssFont = false, jsStyleScript = true }) => (
  <>
    {cssStyleVars ? <StyleThemeColors /> : null}
    {jsStyleScript ? <ScriptThemeValue /> : null}
    {cssFont ? <Fonts /> : null}
  </>
)
