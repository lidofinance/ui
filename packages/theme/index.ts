export * from './cookie-theme-provider'
export * from './theme-provider'
export * from './document-head-contents'
export * from './constants'
export * from './themes'
export * from './hooks'
export * from './types'
export { migrationThemeCookiesToCrossDomainCookiesClientSide } from './utils/cookies-migration'
