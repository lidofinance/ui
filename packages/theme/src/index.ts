export { ThemeProvider } from './provider'
export * from './constants'
export * from './fonts'
export * from './themes'
export * from './types'
