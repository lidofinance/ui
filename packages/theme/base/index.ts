import colors from './colors'
import breakpoints from './breakpoints'
import spacing from './spacing'
import transitions from './transitions'
import typography from './typography'
import borders from './borders'
import shadows from './shadows'

export default {
  ...spacing,
  ...breakpoints,
  ...colors,
  ...transitions,
  ...typography,
  ...borders,
  ...shadows,
}
