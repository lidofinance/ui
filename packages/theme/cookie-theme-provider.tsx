import {
  createContext,
  FC,
  memo,
  PropsWithChildren,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react'
import { ThemeProvider as StyledThemeProvider } from 'styled-components'
import { themeMap } from './themes'
import { getThemeNameFromCookies } from './utils/cookies'
import { initColors } from './utils/document-head-contents'
import { updateGlobalTheme } from './utils/document-head-contents'

export enum ThemeName {
  light = 'light',
  dark = 'dark',
}
export type ThemeNames = keyof typeof ThemeName

export type ThemeContext = {
  toggleTheme: () => void
  themeName: ThemeNames
}

const defaultThemeContext = {} as ThemeContext
export const ThemeToggleContext =
  createContext<ThemeContext>(defaultThemeContext)

export const DEFAULT_THEME_NAME = ThemeName.light

export const prefersDarkThemeMediaQuery =
  typeof window !== 'undefined'
    ? window.matchMedia?.('(prefers-color-scheme: dark)')
    : undefined

// we need to initialize this before react component code if we're using this provider or ThemeProvider
initColors()

/**
 * This is really complicated logic here. Comments will be added on specific lines
 *
 * Cookie theme provider acts differently from common theme provider.
 * 1. it can be nested, BUT it should re-use value provided by itself for optimisation purposes
 * 2. if it is top-level, it is not altering itself, but instead altering theme behavior on HTMLElement level
 * 3. it does not support custom themes, preferring pre-defined themes instead
 * */

export const CookieThemeProvider: FC<
  PropsWithChildren<{
    // Use themeNameParent if you need get cookie in SSR
    initialThemeName?: ThemeName
    overrideThemeName?: ThemeName
  }>
> = memo(
  ({
    children,
    initialThemeName,
    // overrideThemeName is mainly used for storybook
    overrideThemeName,
  }) => {
    const parentTheme = useContext(ThemeToggleContext)
    // we do not want to do nested injections, and we're checking
    // if context we inject in this component is already provided
    const isTopLevelProvider = Object.keys(parentTheme).length === 0
    // we always start with default theme, or, if server wants to provide
    // specific default theme, with server-provided theme to avoid hydration errors
    const [internalThemeName, setThemeName] = useState<ThemeName>(
      initialThemeName || DEFAULT_THEME_NAME,
    )
    // since we're using this component to provide cookie-theme,
    // we eventually want to respect theme provided in cookie, not general theme,
    // so we're picking `parentTheme.themeName` if we have one.
    // If user needs custom theme, not "cookie theme", he should use ThemeProvider instead.
    const themeName = parentTheme.themeName || internalThemeName

    const theme = themeMap[themeName]

    // This logic is really hydrating everything, since we started with server-rendered default prop.
    // It also follows cookie value change.
    useEffect(() => {
      // This logic is useless if we're nested - it is already done in top-level provider
      if (!isTopLevelProvider) {
        return
      }

      const setTheme = () => {
        const systemThemeName = prefersDarkThemeMediaQuery?.matches
          ? ThemeName.dark
          : ThemeName.light
        const themeNameCookie = getThemeNameFromCookies()
        const newThemeName =
          // first, if we have some override (e.g. in Storybook), we respect it
          overrideThemeName ||
          // then, if we have cookie theme, we use theme from cookie
          themeNameCookie ||
          // else, we follow theme we were provided in initialization from server,
          // which means server explicitly wants specific theme by default, not "follow-system"
          initialThemeName ||
          // then, we use media query theme, if available (not all browsers may still support it)
          systemThemeName ||
          // and, finally, falling back to default
          DEFAULT_THEME_NAME
        setThemeName(newThemeName)
        // and when theme is switched, we're setting global-level CSS variable data-attribute
        // to modify CSS vars provided on a top-level
        document.documentElement.dataset.lidoTheme = newThemeName
      }

      // Users may have auto-theme (switching on sunset or schedule or whatever) so we need to listen for this event
      prefersDarkThemeMediaQuery?.addEventListener('change', setTheme)
      setTheme()

      // This code check that the theme cookie was changed on other tab or site (the same second-level domain)
      const checkCookieThemeWasChanged = () => {
        const themeNameCookie = getThemeNameFromCookies()

        if (
          themeNameCookie &&
          (themeNameCookie === ThemeName.dark ||
            themeNameCookie === ThemeName.light)
        ) {
          setThemeName(themeNameCookie)
        }
      }
      window.addEventListener('focus', checkCookieThemeWasChanged)

      return () => {
        window.removeEventListener('focus', checkCookieThemeWasChanged)
      }
    }, [
      initialThemeName,
      isTopLevelProvider,
      overrideThemeName,
      parentTheme,
      theme,
    ])

    const value = useMemo(
      () => ({
        themeName,
        toggleTheme() {
          setThemeName((themeName) => {
            const newThemeName =
              themeName === ThemeName.light ? ThemeName.dark : ThemeName.light
            // note that we're writing cookie theme only on explicit calls,
            // not on "internal theme state" change
            updateGlobalTheme(newThemeName)
            // we do not need to run setTheme here, as effect is triggered
            return newThemeName
          })
        },
      }),
      [themeName],
    )

    if (isTopLevelProvider) {
      // if this provider is top-level we rely on element-theme-script.tsx logic
      // which defines data-lido-theme injection
      return (
        <ThemeToggleContext.Provider value={value}>
          <StyledThemeProvider theme={theme}>{children}</StyledThemeProvider>
        </ThemeToggleContext.Provider>
      )
    } else if (overrideThemeName) {
      return (
        <div style={{ display: 'contents' }} data-lido-theme={themeName}>
          <StyledThemeProvider theme={theme}>{children}</StyledThemeProvider>
        </div>
      )
    } else {
      return <>{children}</>
    }
  },
)

CookieThemeProvider.displayName = 'CookieThemeProvider'
