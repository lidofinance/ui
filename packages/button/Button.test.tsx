import { render } from '@testing-library/react'
import { LightThemeProvider } from '../theme'
import 'jest-styled-components'
import { Button } from './Button'

it('renders correctly', () => {
  const { container } = render(
    <LightThemeProvider>
      <Button>Test</Button>
    </LightThemeProvider>,
  )

  expect(container.firstChild?.firstChild).toMatchSnapshot()
})
