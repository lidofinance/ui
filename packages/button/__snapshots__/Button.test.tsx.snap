// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  box-sizing: border-box;
  margin: 0;
  border: none;
  outline: none;
  white-space: nowrap;
  overflow: hidden;
  position: relative;
  background: transparent;
  font-family: inherit;
  font-weight: 700;
  width: auto;
  line-height: 1em;
  font-size: 14px;
  border-radius: 10px;
  padding: 21px 44px;
  min-width: 120px;
  color: var(--lido-color-primaryContrast);
  background-color: var(--lido-color-primary);
  -webkit-transition: background-color 100ms ease;
  transition: background-color 100ms ease;
}

.c0:before {
  content: '';
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  pointer-events: none;
  border-radius: inherit;
  display: none;
}

.c0:not(:disabled) {
  cursor: pointer;
}

.c0:disabled {
  opacity: 0.5;
}

.c0:not(:disabled):hover,
.c0:focus-visible {
  background-color: var(--lido-color-primaryHover);
}

.c1 {
  position: relative;
  pointer-events: none;
  visibility: visible;
}

<button
  class="c0"
  type="button"
>
  <span
    class="c1"
  >
    Test
  </span>
</button>
`;
